crates/matrix-sdk-common/src/linked_chunk/relational.rs

```diff
@@ -0,0 +1,658 @@
+                    }
+                }
+
+                Update::StartReattachItems | Update::EndReattachItems => { /* nothing */ }
+            }
+        }
+
+        fn insert_chunk(
+            chunks: &mut Vec<ChunkRow>,
+            room_id: &RoomId,
+            previous: &Option<ChunkIdentifier>,
+            new: &ChunkIdentifier,
+            next: &Option<ChunkIdentifier>,
+        ) {
+            // Find the previous chunk, and update its next chunk.
+            if let Some(previous) = previous {
+                let entry_for_previous_chunk = chunks
+                    .iter_mut()
+                    .find(|ChunkRow { room_id: room_id_candidate, chunk, .. }| {
+                        room_id == room_id_candidate && chunk == previous
+                    })
+                    .expect("Previous chunk should be present");
+
+                // Insert the chunk.
+                entry_for_previous_chunk.next_chunk = Some(*new);
+            }
+
+            // Find the next chunk, and update its previous chunk.
+            if let Some(next) = next {
+                let entry_for_next_chunk = chunks
+                    .iter_mut()
+                    .find(|ChunkRow { room_id: room_id_candidate, chunk, .. }| {
+                        room_id == room_id_candidate && chunk == next
+                    })
+                    .expect("Next chunk should be present");
+
+                // Insert the chunk.
+                entry_for_next_chunk.previous_chunk = Some(*new);
+            }
+
+            // Insert the chunk.
+            chunks.push(ChunkRow {
+                room_id: room_id.to_owned(),
+                previous_chunk: *previous,
+                chunk: *new,
+                next_chunk: *next,
+            });
+        }
+
+        fn remove_chunk(
+            chunks: &mut Vec<ChunkRow>,
+            room_id: &RoomId,
+            chunk_to_remove: &ChunkIdentifier,
+        ) {
+            let entry_nth_to_remove = chunks
+                .iter()
+                .enumerate()
+                .find_map(|(nth, ChunkRow { room_id: room_id_candidate, chunk, .. })| {
+                    (room_id == room_id_candidate && chunk == chunk_to_remove).then_some(nth)
+                })
+                .expect("Remove an unknown chunk");
+
+            let ChunkRow { room_id, previous_chunk: previous, next_chunk: next, .. } =
+                chunks.remove(entry_nth_to_remove);
+
+            // Find the previous chunk, and update its next chunk.
+            if let Some(previous) = previous {
+                let entry_for_previous_chunk = chunks
+                    .iter_mut()
+                    .find(|ChunkRow { room_id: room_id_candidate, chunk, .. }| {
+                        &room_id == room_id_candidate && *chunk == previous
+                    })
+                    .expect("Previous chunk should be present");
+
+                // Insert the chunk.
```

Comment:
`Insert` isn't the right action, maybe `Link` instead?

```suggestion
                // Link the chunk.
```